ndbench-es-plugins/src/main/java/com/netflix/ndbench/plugin/es/ConstantStepWiseRateIncreaser.java

```diff
@@ -0,0 +1,150 @@
+        }
+        if (rampPeriodMillisecs / incrementIntervalMillisecs > MAX_STEPS) {
+            throw new IllegalArgumentException(
+                    "rampPeriodMillisecs / incrementIntervalMillisecs should not exceed MAX_STEPS (" + MAX_STEPS + ")");
+        }
+
+
+        final ArrayList<Cell> tmpClockTimeToRateTable = new ArrayList<Cell>();
+
+        int numSteps = rampPeriodMillisecs / incrementIntervalMillisecs;
+        double spread = (finalRate - initRate) * 1.0;
+        double rateIncrementPerStep = spread / numSteps;
+
+        for (int i = 0; i <= numSteps; i++) {
+            tmpClockTimeToRateTable.add(
+                    new Cell(
+                            i * incrementIntervalMillisecs,
+                            initRate + (i * rateIncrementPerStep)
+                    )
+            );
+        }
+
+        // initialize high values sentinel cell
+        tmpClockTimeToRateTable.add(new Cell(Long.MAX_VALUE, 0 /* dummy never used for rate */));
+
+        indexOfLastInTable = numSteps + 1;
+        assert (tmpClockTimeToRateTable.size() == indexOfLastInTable + 1);
+        assert (tmpClockTimeToRateTable.get(numSteps).time == rampPeriodMillisecs);
+        assert (tmpClockTimeToRateTable.get(numSteps).rate == finalRate);
+
+
+        ImmutableList.Builder<Cell> cellBuilder = new ImmutableList.Builder<>();
+        clockTimeToRateTable = cellBuilder.addAll(tmpClockTimeToRateTable).build();
+    }
+
+    /**
+     * Build internal array mapping relative-from-zero clock time values to the rate that will be returned once system
+     * clock exceeds time value in a particular cell (but is less than the time value in the immediately
+     * succeeding cell);
+     * <p>
+     * Ignore the  possible race condition that arises if multiple threads call this method at around the same time
+     * when an instance is an uninitialized state. Two (or maybe more) arrays will be concurrently built, but the
+     * system clock time to rate mappings will be close enough that it won't really matter -- auto-tuning behavior
+     * will work as expected.
+     * <p>
+     * TODO -- put this comment somewhere else
+     */
+
+    public double getRateForGivenClockTime(long baseReferenceTime, long clockTime) {
+        if (baseReferenceTime > clockTime) {
+            throw new IllegalArgumentException(
+                    "specified baseReferenceTime ("
+                            + baseReferenceTime + ") is greater than clockTime (" + clockTime + ")");
+
+        }
+
+        long desiredClockTimeRelativizedToTimeZero = clockTime - baseReferenceTime;
+        double retval = -1;
+
+        // Find the first cell 'j' whose immediate successor cell (j+1) has a time that exceeds desiredClockTime,
+        // then return rate for cell 'j'
+        //
+        for (int i = 0; i <= indexOfLastInTable; i++) {
+            long timeCeilingFromSuccessor = clockTimeToRateTable.get(i + 1).time;
```

Comment:
Could you double check you are not running out of bounds in clockTimeToRateTable here? indexOfLastInTable is supposed to be pointing to the last element, and you are getting `i + 1`.

Reply:
i unit tested this pretty exhaustively you still found a bug  ;^0  
the bug would have never resulted in a blow up, because the sentinel value (with LONG.highvalues) is always higher than any relativized time you woudl pass in.    But <=  is  confusing to those who read the code in detail like you.. So i changed to "<"      /  thanks